### PR TITLE
[JENKINS-61751] let :disabled and :deleted at the same time

### PR DESCRIPTION
**Pull request by ivarmu**
Asked for a new capability derived from https://github.com/jenkinsci/active-choices-plugin/pull/25#issuecomment-601925184.

Allow to disable and select any (select option|radio|checkbox) with :disable and/or :selected

### CODE REVIEW

**Comment by ivarmu** [comment]
Here is an image for the documentation update. Location marked by **IMAGE HERE**
image

**Comment by ivarmu** [comment]
Yes, this would happen... I think it'll be easy to rewrite the code to
detect the suffices... I'll take a look tomorrow.

**Comment by SciMan72** [comment]
Thanks a lot for your effort and the fast support. I'm looking forward for the new version.

**Comment by ivarmu** [comment]
image
New image version for the document showing the new modifications.

**Comment by ivarmu** [comment]
@kinow how can I relaunch the tests? They are failed because of unavailable Jenkins nodes:

> [2020-03-23T10:40:41.598Z] Still waiting to schedule task
> [2020-03-23T10:40:41.598Z] ‘EC2 (aws) - Windows 2019 (i-0bb38e192d913b7da)’ is offline
> [2020-03-23T10:55:22.657Z] EC2 (aws) - Ubuntu 18.04 LTS (i-05a915b10f8fa9f24) was marked offline: Connection was broken: java.io.EOFException
> [2020-03-23T10:55:22.657Z] 	at java.base/java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2818)

**Comment by kinow** [comment]
@ivaivalous I tried closing and re-openning as that works with Travis CI and GitHub actions, but didn't work here. I've also logged-in to https://ci.jenkins.io/, but no luck there. No button to reload it. You can try an empty commit perhaps? (that one with `--allow-empty` I think)

**Comment by ivarmu** [comment]
> @ivaivalous I tried closing and re-openning as that works with Travis CI and GitHub actions, but didn't work here. I've also logged-in to https://ci.jenkins.io/, but no luck there. No button to reload it. You can try an empty commit perhaps? (that one with `--allow-empty` I think)

It worked, but the tests won't be successfull as there's no available windows nodes to run them on:

> 21:23:37  Still waiting to schedule task
> 21:23:37  ‘EC2 (aws) - Windows 2019 (i-018533cfd61ab6cf4)’ is offline

https://ci.jenkins.io/computer/EC2%20(aws)%20-%20Windows%202019%20(i-018533cfd61ab6cf4)/

**Comment by ivarmu** [comment]
> @ivaivalous I tried closing and re-openning as that works with Travis CI and GitHub actions, but didn't work here. I've also logged-in to https://ci.jenkins.io/, but no luck there. No button to reload it. You can try an empty commit perhaps? (that one with `--allow-empty` I think)

It worked! So all the tests are fine right now :+1:

**Comment by kinow** [comment]
Tested with a single parameter first, and it worked as expected.

image

image

Then tested adding a reactive parameter, and it also worked fine. +1

File: src/main/resources/org/biouno/unochoice/common/selectContent.jelly [hunk 1]
```diff
@@ -25,8 +25,17 @@
...
+                <j:arg type="java.lang.Object" value="${escapedValue}" />
+            </j:invokeStatic>
+            <j:invokeStatic className="org.biouno.unochoice.util.Utils" method="escapeDisabled" var="escapedAndDisabledKey">
+                <j:arg type="java.lang.Object" value="${escapedKey}" />
+            </j:invokeStatic>
+            <option selected="${selected}" disabled="${disabled}" value="${escapedAndDisabledKey}">${escapedAndDisabledValue}</option>
+        </j:when>
         <j:when test="${selected}">
-            <f:option selected="${selected}" value="${escapedKey}">${escapedValue}</f:option>
+            <option selected="${selected}" value="${escapedKey}">${escapedValue}</option>
```
[the remark this answers is not in the document]
Comment (ivarmu) [reply src/main/resources/org/biouno/unochoice/common/selectContent.jelly]: I compared the generated codes and there was no difference... Jelly ```option``` didn't let to specify ```disabled``` in any way, so I changed to HTML option to make ```disabled``` possible.

**Comment by agabrys** [comment]
Hello,
Could you update these lines to?

https://github.com/ivarmu/active-choices-plugin/blob/04cf5ec8d39d63c907879cb91bcd498fe5d837a7/src/main/java/org/biouno/unochoice/AbstractScriptableParameter.java#L293-L294

I think this:
```
if (valueText.endsWith(":selected")) {
    defaultValues.add(valueText.substring(0, valueText.length() - ":selected".length()));
}
```
should be replaced by:
```
if (Utils.isSelected(valueText)) {
    defaultValues.add(Utils.escapeSelected(Utils.escapeDisabled(valueText)));
}
```
It would be nice to add `Utils#escape` and make it shorter:
```
if (Utils.isSelected(valueText)) {
    defaultValues.add(Utils.escape(valueText));
}
```

**Comment by ivarmu** [comment]
@agabrys looks great! I'm taking a look as soon as possible... may be more charles line that' one

**Comment by ivarmu** [comment]
> Could you update these lines to?
> 
> https://github.com/jenkinsci/active-choices-plugin/blob/04cf5ec8d39d63c907879cb91bcd498fe5d837a7/src/main/java/org/biouno/unochoice/AbstractScriptableParameter.java#L293-L295
> 
> I think they should be replaced by:
> 
> ```
> if (Utils.isSelected(valueText)) {
>     defaultValues.add(Utils.escapeSelected(Utils.escapeDisabled(valueText)));
> }
> ```
> 
> It would be nice to add `Utils#escape` and make it shorter:
> 
> ```
> if (Utils.isSelected(valueText)) {
>     defaultValues.add(Utils.escape(valueText));
> }
> ```

So I've updated the code as you requested, but not used the escapeDisabled function, as that section of code only treats ':selected' suffix.

I'm not sure when this file is used... and if it should treat ':disabled' suffix as well... 

Any inputs on this will be appreciated!

**Comment by agabrys** [comment]
I'm the author of this code:
https://github.com/jenkinsci/active-choices-plugin/blob/04cf5ec8d39d63c907879cb91bcd498fe5d837a7/src/main/java/org/biouno/unochoice/AbstractScriptableParameter.java#L293-L295

I didn't know about the `Utils` class and that `:disabled` could be added. The answer is - yes. The `:disabled` should be also trimmed. The clear value without any suffixes should be added into `defaultValues`.

Feel free to extend this test class: https://github.com/jenkinsci/active-choices-plugin/blob/04cf5ec8d39d63c907879cb91bcd498fe5d837a7/src/test/java/org/biouno/unochoice/issue61068/TestDefaultValuesOnParamBuildPage.java

I would at least propose:
```
@Test
public void testAllSuffixesAreTrimmed() {
    ChoiceParameter parameter = createChoiceParameter(
            ChoiceParameter.PARAMETER_TYPE_CHECK_BOX,
            "return ['A:slected:disabled', 'B:disabled:selected', 'C', 'D:selected']"
    );

    ParameterValue parameterValue = parameter.getDefaultParameterValue();

    assertEquals("Invalid parameter name!", PARAMETER_NAME, parameterValue.getName());
    assertEquals("Invalid parameter value!", "A,B,D", parameterValue.getValue());
}
```

**Comment by ivarmu** [comment]
> I'm the author of this code:
> https://github.com/jenkinsci/active-choices-plugin/blob/04cf5ec8d39d63c907879cb91bcd498fe5d837a7/src/main/java/org/biouno/unochoice/AbstractScriptableParameter.java#L293-L295
> 
> I didn't know about the `Utils` class and that `:disabled` could be added. The answer is - yes. The `:disabled` should be also trimmed. The clear value without any suffixes should be added into `defaultValues`.
> 
> Feel free to extend this test class: https://github.com/jenkinsci/active-choices-plugin/blob/04cf5ec8d39d63c907879cb91bcd498fe5d837a7/src/test/java/org/biouno/unochoice/issue61068/TestDefaultValuesOnParamBuildPage.java
> 
> I would at least propose:
> 
> ```
> @Test
> public void testAllSuffixesAreTrimmed() {
>     ChoiceParameter parameter = createChoiceParameter(
>             ChoiceParameter.PARAMETER_TYPE_CHECK_BOX,
>             "return ['A:selected:disabled', 'B:disabled:selected', 'C', 'D:selected']"
>     );
> 
>     ParameterValue parameterValue = parameter.getDefaultParameterValue();
> 
>     assertEquals("Invalid parameter name!", PARAMETER_NAME, parameterValue.getName());
>     assertEquals("Invalid parameter value!", "A,B,D", parameterValue.getValue());
> }
> ```

Great! I'm doing the proper coding tonight if all goes well. Thanks a lot @agabrys!!!

**Comment by ivarmu** [comment]
Hi guys, when dou you expect to merge this PR and release the new version?

**Comment by kinow** [comment]
Sorry for the delay @ivarmu . Merged! Thanks!

Yesterday I started working on the updates to pom.xml to prepare a new release. Will cut it either this week or this next weekend if time allows.

**Comment by ivarmu** [comment]
@kinow  I see the image is missing at the current version of the documentation... I can still read the IMAGE HERE placeholder for the provided image. The correct version of the image follows. I couldn't find the way to upload the image the right way by myself :relieved:

**Comment by ivarmu** [comment]
@kinow  that's the correct version of the image. :blush:

**Comment by ivarmu** [comment]
Sorry @kinow https://github.com/jenkinsci/active-choices-plugin/pull/33#issuecomment-602518924

**Comment by kinow** [comment]
Hi @ivarmu , missed that one. Updated README. Thanks!!

**Pull request merged**